server/src/main/java/org/elasticsearch/cluster/coordination/ElectionScheduler.java

```diff
@@ -0,0 +1,246 @@
+        if (electionMaxRetryInterval.millis() < electionMinRetryInterval.millis() + 100) {
+            throw new IllegalArgumentException(validationExceptionMessage(
+                electionMinRetryInterval.toString(), electionMaxRetryInterval.toString()));
+        }
+    }
+
+    static String validationExceptionMessage(final String electionMinRetryInterval, final String electionMaxRetryInterval) {
+        return new ParameterizedMessage(
+            "Invalid election retry intervals: [{}] is [{}] and [{}] is [{}], but [{}] should be at least 100ms longer than [{}]",
+            ELECTION_MIN_RETRY_INTERVAL_SETTING_KEY, electionMinRetryInterval,
+            ELECTION_MAX_RETRY_INTERVAL_SETTING_KEY, electionMaxRetryInterval,
+            ELECTION_MAX_RETRY_INTERVAL_SETTING_KEY, ELECTION_MIN_RETRY_INTERVAL_SETTING_KEY).getFormattedMessage();
+    }
+
+    private Object currentScheduler; // only care about its identity
+    private final TimeValue electionMinRetryInterval;
+    private final TimeValue electionMaxRetryInterval;
+    private final Object mutex = new Object();
+    private final ThreadPool threadPool;
+    private final Random random;
+
+    private long currentDelayMillis;
+
+    ElectionScheduler(Settings settings, Random random, ThreadPool threadPool) {
+        super(settings);
+
+        this.random = random;
+        this.threadPool = threadPool;
+
+        synchronized (mutex) {
+            electionMinRetryInterval = ELECTION_MIN_RETRY_INTERVAL_SETTING.get(settings);
+            electionMaxRetryInterval = ELECTION_MAX_RETRY_INTERVAL_SETTING.get(settings);
+        }
+    }
+
+    public void start() {
+        logger.trace("starting");
+        final Runnable newScheduler;
+        synchronized (mutex) {
+            currentDelayMillis = electionMinRetryInterval.millis();
+            assert currentScheduler == null;
+            currentScheduler = newScheduler = new Runnable() {
+                @Override
+                public String toString() {
+                    return "ElectionScheduler: next election scheduler";
+                }
+
+                private boolean isCurrentScheduler() {
+                    return this == currentScheduler;
+                }
+
+                private void scheduleNextElection() {
+                    final long delay;
+                    synchronized (mutex) {
+                        if (isCurrentScheduler() == false) {
+                            logger.trace("not scheduling election");
+                            return;
+                        }
+                        currentDelayMillis = Math.min(electionMaxRetryInterval.getMillis(),
+                            currentDelayMillis + electionMinRetryInterval.getMillis());
+
+                        delay = randomLongBetween(electionMinRetryInterval.getMillis(), currentDelayMillis + 1);
+                    }
+                    logger.trace("scheduling election after delay of [{}ms]", delay);
```

Comment:
maybe `with delay` instead of `after delay` (after sounds as if the delay has already passed)